### PR TITLE
Add LogitProcessor interface for pre-sampling logit transforms (#19517)

### PR DESCRIPTION
**Pull request by kirklandsign**
Summary:

Introduces a `LogitProcessor` abstract interface that lets callers mutate
logits in place between the model forward pass and the sampler. Enables
grammar-constrained decoding, logit biasing, repetition penalties, and
similar transforms without touching the core generation loop.

Interface (`extension/llm/sampler/logit_processor.h`, ~15 lines):
- Single virtual method `process(::executorch::aten::Tensor logits)` that
  returns `Error::Ok` or aborts the chain on a non-Ok return.
- Tensor passed by value (handle-typed ATen idiom; mutations propagate
  through the shared underlying buffer).
- Each implementation declares its own dtype expectations -- the chain
  runner does not cast or copy the tensor. Typical implementations check
  `logits.scalar_type()` and either dispatch to a kernel or return
  InvalidArgument.
- Tensor shape contract (rank 2 = `[batch, vocab]`, rank 3 =
  `[batch, seq, vocab]` advanced to last sequence position) mirrors
  `sample_from_logits`.

Wiring (`extension/llm/runner/text_token_generator.h`):
- New public methods `add_logit_processor`, `clear_logit_processors`,
  `num_logit_processors`.
- Inside `generate()`, between `step()` and `logits_to_token()`, the loop
  invokes each registered processor:
    for (auto& processor : logit_processors_) {
      ET_CHECK_OK_OR_RETURN_ERROR(processor->process(logits_tensor));
    }
- Empty chain is the existing fast path; no behavior change for callers
  that don't register processors.

Configure processors before calling `generate()` -- concurrent
modification during generation is not safe.

Reviewed By: Gasoonjia

Differential Revision: D104767967

cc @larryliu0820 @mergennachin @cccclai @helunwencser @jackzhxng

### CODE REVIEW

**Comment by kirklandsign** [comment]
@claude review

**Comment by kirklandsign** [comment]
@claude review

**Review by Gasoonjia — APPROVED** [review]
